torch/csrc/distributed/rpc/rpc_with_autograd.cpp

```diff
@@ -83,64 +89,64 @@ Message RpcWithAutograd::toMessage() && {
 
 std::unique_ptr<RpcWithAutograd> RpcWithAutograd::fromMessage(
     const Message& message) {
-  MessageType originalMessageType = message.type();
   TORCH_INTERNAL_ASSERT(
-      MessageType::MESSAGE_WITH_AUTOGRAD_REQ == originalMessageType ||
-      MessageType::MESSAGE_WITH_AUTOGRAD_RESP == originalMessageType);
+      MessageType::MESSAGE_WITH_AUTOGRAD_REQ == message.type() ||
+      MessageType::MESSAGE_WITH_AUTOGRAD_RESP == message.type());
 
-  std::vector<torch::Tensor> tensors = message.tensors();
-  int64_t messageId = message.id();
-  // Decode message type, autograd context id and autograd message id.
+  // Decode message type, autograd context id, autograd message id and worker
+  // id.
   auto payload = message.payload();
-  TORCH_INTERNAL_ASSERT(payload.size() > kAutogradMessageSize);
 
-  int64_t autogradContextId, autogradMessageId;
-  // autograd message id.
+  // Read the autograd payload remove it from the payload.
+  int64_t autogradPayLoadSize;
   size_t indexToRead = payload.size() - sizeof(int64_t);
   TORCH_INTERNAL_ASSERT(indexToRead >= 0);
   torch::utils::THP_decodeInt64Buffer(
-      &autogradMessageId,
-      reinterpret_cast<uint8_t*>(payload.data()) + indexToRead,
-      torch::utils::THPByteOrder::THP_BIG_ENDIAN,
-      1);
-
-  // autograd context id.
-  indexToRead -= sizeof(int64_t);
-  TORCH_INTERNAL_ASSERT(indexToRead >= 0);
-  torch::utils::THP_decodeInt64Buffer(
-      &autogradContextId,
+      &autogradPayLoadSize,
       reinterpret_cast<uint8_t*>(payload.data()) + indexToRead,
       torch::utils::THPByteOrder::THP_BIG_ENDIAN,
       1);
-
-  // message type.
-  indexToRead -= 1;
-  TORCH_INTERNAL_ASSERT(indexToRead >= 0);
+  payload.resize(indexToRead);
+
+  // Now read the entire autograd payload and unpickle.
+  TORCH_INTERNAL_ASSERT(payload.size() > autogradPayLoadSize)
+  auto autogradPayLoadBegin =
+      static_cast<const char*>(message.payload().data()) + payload.size() -
+      autogradPayLoadSize;
+  std::vector<torch::Tensor> tensorTable;
+  IValue tuple = jit::unpickle(
```

Comment:
This refactor of `fromMessage` is picked from https://github.com/pytorch/pytorch/pull/27022, and is needed so that the `RpcWithAutograd` can be constructed correctly (i.e. it requires a workerId).